test/TestAppUtils/VisualTreeDumper.cs

```diff
@@ -0,0 +1,184 @@
+                _indent = 0;
+                _filter = filter;
+                _translator = translator;
+            }
+            public void EndVisitNode(DependencyObject obj)
+            {
+                _indent--;
+                AddPadding(_indent);
+                _sb.AppendLine("</Element>");
+            }
+
+            public void BeginVisitNode(DependencyObject obj)
+            {
+                AddPadding(_indent);
+                _sb.AppendLine(String.Format("<Element Type=\"{0}\">", obj.GetType().FullName));
+                _indent++;
+            }
+
+            public override String ToString()
+            {
+                return _sb.ToString();
+            }
+
+            public bool ShouldVisitNode(DependencyObject node)
+            {
+                return node != null && _filter.ShouldLogElement(node.GetType().FullName);
+            }
+
+            public bool ShouldVisitPropertiesForNode(DependencyObject node)
+            {
+                return (node as UIElement) != null && _filter.ShouldLogElement(node.GetType().FullName);
+            }
+
+            public bool ShouldVisitProperty(PropertyInfo propertyInfo)
+            {
+                return _filter.ShouldLogProperty(propertyInfo.Name);
+            }
+            public void VisitProperty(string propertyName, object value)
+            {
+                var v = _translator.PropertyValueToString(propertyName, value);
+                if (!_filter.IsKnownProperty(propertyName, v))
+                {
+                    AddPadding(_indent + 1);
+                    _sb.AppendLine(String.Format("<Property Name=\"{0}\" Value=\"{1}\" />", propertyName, v));
+                }
+            }
+
+            private void AddPadding(int numOfSpace)
+            {
+                _sb.Append("".PadRight(numOfSpace));
+            }
+        }
+
+        public static String DumpToXML(DependencyObject root, IPropertyValueTranslator translator, IFilter filter)
+        {
+
+            Visitor visitor = new Visitor(filter ?? new DefaultFilter(), translator ?? new DefaultPropertyValueTranslator());
+            WalkThroughTree(root, visitor);
+            return visitor.ToString();
+        }
+
+        private static void WalkThroughProperties(DependencyObject node, IVisitor visitor)
+        {
+            if (visitor.ShouldVisitPropertiesForNode(node))
+            {
+                var properties = node.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance);
+                foreach (var property in properties)
+                {
+                    if (visitor.ShouldVisitProperty(property))
+                    {
+                        Object value = null;
+
+                        try
+                        {
+                            value = property.GetValue(node, null);
+                        }
+                        catch (Exception)
+                        {
+                            value = "Exception";
```

Comment:
Wait, the string "Exception" becomes the property value? Not the type/message? What types of exceptions were you seeing?

Reply:
Maybe it's introduced by compiler. when running on RS5 machine, Things like Shadow would be in the property list, which is actually introduced since 19H1.

Reply:
I don't understand your comment here. Why are you using the string "Exception" here instead of something more helpful?

Reply:
I don't want to make it too complex,. the exception would look like this: "Exception when read " + property.Name

---
In reply to: [262285261](https://github.com/Microsoft/microsoft-ui-xaml/pull/126#discussion_r262285261) [](ancestors = 262285261)